<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>

<body>
    <script>

        function increase(number) {
            const promise = new Promise((resolve, reject) => {
                setTimeout(() => {
                    const result = number + 10;
                    if (result > 50) {
                        const error = new Error('NumberIsTooBig');
                        return reject(error);
                    }
                    resolve(result);
                }, 1000);
            });
            return promise;
        }

        /* 
            await의 2가지 기능
            1. await가 달린 함수의 결과인(promise에 담긴 결과(resolve ('결과'))를 반환해준다.
            2. await가 달린 비동기처리는 동기식으로 동작한다.
        */
        async function run() {
            try {
                let result = await increase(0);
                console.log(result);

                result = await increase(result);
                console.log(result);

                result = await increase(result);
                console.log(result);

                result = await increase(result);
                console.log(result);

                result = await increase(result);
                console.log(result);

                result = await increase(result);
                console.log(result);

            } catch (error) {
                console.log(error, '위 에러가 발생했습니다');
            }
        }

        run();
    </script>
</body>

</html>